File: android/src/main/java/com/margelo/nitro/nitrometamask/HybridMetamaskConnector.kt
package com.margelo.nitro.nitrometamask

import com.margelo.nitro.Promise
import io.metamask.androidsdk.Ethereum
import io.metamask.androidsdk.Result
import io.metamask.androidsdk.DappMetadata
import io.metamask.androidsdk.SDKOptions
import com.facebook.react.bridge.ReactApplicationContext
import kotlinx.coroutines.suspendCoroutine

class HybridMetamaskConnector : HybridMetamaskConnectorSpec() {
    companion object {
        @Volatile
        private var reactContext: ReactApplicationContext? = null
        
        fun setReactContext(context: ReactApplicationContext) {
            reactContext = context
        }
    }
    
    // Initialize Ethereum SDK with Context, DappMetadata, and SDKOptions
    // Based on: https://github.com/MetaMask/metamask-android-sdk
    // The SDK requires a Context for initialization
    private val ethereum: Ethereum by lazy {
        val context = reactContext?.applicationContext
            ?: throw IllegalStateException("ReactApplicationContext not initialized. Make sure NitroMetamaskPackage is properly registered.")
        
        val dappMetadata = DappMetadata(
            name = "Nitro MetaMask Connector",
            url = "https://novastera.com"
        )
        val sdkOptions = SDKOptions()
        
        Ethereum(context, dappMetadata, sdkOptions)
    }

    override fun connect(): Promise<ConnectResult> {
        // Use Promise.async with coroutines for best practice in Nitro modules
        // Reference: https://nitro.margelo.com/docs/types/promises
        return Promise.async {
            // Convert callback-based connect() to suspend function using suspendCoroutine
            val result = suspendCoroutine<Result> { continuation ->
                ethereum.connect { callbackResult ->
                    continuation.resume(callbackResult)
                }
            }
            
            when (result) {
                is Result.Success.Item -> {
                    // After successful connection, get account info from SDK
                    val address = ethereum.selectedAddress
                        ?: throw IllegalStateException("MetaMask SDK returned no address after connection")
                    val chainId = ethereum.chainId
                        ?: throw IllegalStateException("MetaMask SDK returned no chainId after connection")
                    
                    ConnectResult(
                        address = address,
                        chainId = chainId.toString()
                    )
                }
                is Result.Error -> {
                    throw Exception(result.error.message ?: "Failed to connect to MetaMask")
                }
                else -> {
                    throw IllegalStateException("Unexpected result type from MetaMask connect")
                }
            }
        }
    }

    override fun signMessage(message: String): Promise<String> {
        // Use Promise.async with coroutines for best practice in Nitro modules
        // Reference: https://nitro.margelo.com/docs/types/promises
        return Promise.async {
            // Use the convenience method connectSign() which connects and signs in one call
            // Based on SDK docs: ethereum.connectSign(message) returns Result synchronously
            // Reference: https://github.com/MetaMask/metamask-android-sdk
            when (val result = ethereum.connectSign(message)) {
                is Result.Success.Item -> {
                    // Extract signature from result
                    result.value as? String
                        ?: throw IllegalStateException("Invalid signature response format")
                }
                is Result.Error -> {
                    throw Exception(result.error.message ?: "Failed to sign message")
                }
                else -> {
                    throw IllegalStateException("Unexpected result type from MetaMask signMessage")
                }
            }
        }
    }
}

File: ios/HybridMetamaskConnector.swift
import Foundation
import MetaMaskSDK
import NitroModules

class HybridMetamaskConnector: HybridMetamaskConnectorSpec {
  private let sdk = MetaMaskSDK.shared

  func connect() -> Promise<ConnectResult> {
    // Use Promise.async with Swift async/await for best practice in Nitro modules
    // Reference: https://nitro.margelo.com/docs/types/promises
    return Promise.async {
      // Based on MetaMask iOS SDK docs: let connectResult = await metamaskSDK.connect()
      // Reference: https://github.com/MetaMask/metamask-ios-sdk
      let connectResult = try await self.sdk.connect()
      
      switch connectResult {
      case .success(let value):
        // After successful connection, get account info from SDK
        // Note: sdk.account is a String (address), not an object
        // Reference: https://raw.githubusercontent.com/MetaMask/metamask-ios-sdk/924d91bb3e98a5383c3082d6d5ba3ddac9e1c565/README.md
        guard let address = self.sdk.account, !address.isEmpty else {
          throw NSError(domain: "MetamaskConnector", code: -1, userInfo: [NSLocalizedDescriptionKey: "MetaMask SDK returned no address after connection"])
        }
        guard let chainId = self.sdk.chainId, !chainId.isEmpty else {
          throw NSError(domain: "MetamaskConnector", code: -1, userInfo: [NSLocalizedDescriptionKey: "MetaMask SDK returned no chainId after connection"])
        }
        return ConnectResult(address: address, chainId: chainId)
      case .failure(let error):
        throw error
      }
    }
  }

  func signMessage(message: String) -> Promise<String> {
    // Use Promise.async with Swift async/await for best practice in Nitro modules
    // Reference: https://nitro.margelo.com/docs/types/promises
    return Promise.async {
      // Use the convenience method connectAndSign() which connects and signs in one call
      // This is equivalent to Android's connectSign() method
      // Reference: https://raw.githubusercontent.com/MetaMask/metamask-ios-sdk/924d91bb3e98a5383c3082d6d5ba3ddac9e1c565/README.md
      // Example: https://raw.githubusercontent.com/MetaMask/metamask-ios-sdk/924d91bb3e98a5383c3082d6d5ba3ddac9e1c565/Example/metamask-ios-sdk/SignView.swift
      let connectSignResult = try await self.sdk.connectAndSign(message: message)
      
      switch connectSignResult {
      case .success(let signature):
        return signature
      case .failure(let error):
        throw error
      }
    }
  }
}

